<a href="https://colab.research.google.com/github/masterPLIA2425/crei-practica4-Rebecago1/blob/main/practica4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Clasificación, resumen y extracción de la información**

# **Práctica 4**

En esta práctica vamos a programar el método de clasificación conocido como *Naive Bayes*. Lo haremos en dos versiones: con vectorización por medio de sows y con vectorización por medio de bows.

Vamos a suponer que tenemos un *dataset* anotado con la siguiente estructura: se trata de un diccionario que a cada etiqueta/clase/categoría le asocia una lista de textos (= cadenas de caracteres).

El ejemplo que vamos a utilizar tiene solo dos etiquetas: "deportes" y "arte", y los ejemplos anotados son los siguientes.

In [ ]:
dataSetEjemplo=dict()

In [ ]:
dataSetEjemplo["deportes"] = ["Con un segundo gol, el centrocampista del PSG rescata del asedio helvético a la selección, que jugó más de una hora con uno menos por la expulsión de un jugador. El tercer gol fue anulado.","El jugador asegura que es inocente y pide perdón al club por incumplir el rigor de los horarios y no poder ir a la selección","La plata olímpica y el oro europeo de la selección femenina impulsan la idea de crear futuras competiciones nacionales"]

In [ ]:
dataSetEjemplo["deportes"]

['Con un segundo gol, el centrocampista del PSG rescata del asedio helvético a la selección, que jugó más de una hora con uno menos por la expulsión de un jugador. El tercer gol fue anulado.',
 'El jugador asegura que es inocente y pide perdón al club por incumplir el rigor de los horarios y no poder ir a la selección',
 'La plata olímpica y el oro europeo de la selección femenina impulsan la idea de crear futuras competiciones nacionales']

In [ ]:
dataSetEjemplo["arte"] = ["El autor de Soldados de Salamina se convierte en el quinto escritor español entrevistado en 70 años por la revista neoyorquina que orienta el canon contemporáneo. Se trata de un reconocido autor de novelas y de no ficción.","Una exposición en Berlín rinde homenaje a las mujeres de diversas clases e ideologías que se jugaron la vida para hacer frente al nacionalsocialismo desde la resistencia. La exposición se prolongará durante dos meses. Una exposición en Berlín rinde homenaje a las mujeres de diversas clases e ideologías que se jugaron la vida para hacer frente al nacionalsocialismo desde la resistencia. La exposición se prolongará durante dos meses. La selección fue decidida por un comité.","Poetas consagrados y las voces más jóvenes de la poesía española reflexionan sobre la obra más icónica de Rafael Alberti, escrita cuando tenía 22 años y por la que consiguió el Premio Nacional de poesía en 1924."]

In [ ]:
dataSetEjemplo["arte"]

['El autor de Soldados de Salamina se convierte en el quinto escritor español entrevistado en 70 años por la revista neoyorquina que orienta el canon contemporáneo. Se trata de un reconocido autor de novelas y de no ficción.',
 'Una exposición en Berlín rinde homenaje a las mujeres de diversas clases e ideologías que se jugaron la vida para hacer frente al nacionalsocialismo desde la resistencia. La exposición se prolongará durante dos meses. Una exposición en Berlín rinde homenaje a las mujeres de diversas clases e ideologías que se jugaron la vida para hacer frente al nacionalsocialismo desde la resistencia. La exposición se prolongará durante dos meses. La selección fue decidida por un comité.',
 'Poetas consagrados y las voces más jóvenes de la poesía española reflexionan sobre la obra más icónica de Rafael Alberti, escrita cuando tenía 22 años y por la que consiguió el Premio Nacional de poesía en 1924.']

Vamos a definir también dos ejemplos, que intentaremos clasificar más adelante.

In [ ]:
test1 = "La selección impide que el jugador contabilice el gol"

In [ ]:
test2 = "La exposición de poesía cuenta con una selección de obras de un autor catalán"

Tanto para la vectorización SoW como para la BoW necesitaremos un vocabulario. En el ejemplo será el siguiente:

In [ ]:
vocabularioEjemplo = ["gol", "selección", "jugador", "club", "plata", "oro", "autor", "exposición", "poesía", "resistencia"]

En primer lugar vamos a pasar todos los textos al modelo SoW (que es como decir que usaremos un modelo de documento de Bernouilli).

**Ejercicio 1**. Retomar de las prácticas anteriores todas las funciones necesarias para poder programar sows (con diccionarios). Escribirlas en una única celda, una tras otra, y ejecutar la celda.

In [ ]:
import nltk
nltk.download('punkt_tab')
from nltk.tokenize import word_tokenize

def coincide(cadena1,cadena2):
  return(cadena1 == cadena2)

def aparece (cadena, listaCadenas):
    for palabra in listaCadenas:
        if coincide(cadena,palabra):
            return True
    return False

def apareceEnTexto(cadena,texto):
    return aparece(cadena, word_tokenize(texto))

def sow_dic (texto, vocabulario):
    dict={}
    for cadena in vocabulario:
      if apareceEnTexto(cadena,texto):
        dict[cadena]=1
    return dict


[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


**Las funciones están encadenadas, cada una depende de la anterior para realizar la representación *Sow* mediante un diccionario en el que la palabra del vocabulario que aparece en el texto es la clave y su valor, 1, que indica su presencia.
De este modo, *sow_dic* se basa en *apareceEnTexto()* para determinar si se debe agregar una palabra al diccionario con un 1; *apareceEnTexto()*, a su vez, se fundamenta en *aparece()* y *aparece()* necesita a *coincide()* para identificar la palabra en sí.
Con esto se demuestra, por un lado, la modularidad del programa, por otro la importancia de verificar que las funciones estén bien definidas para que no se concatenen errores.**

In [ ]:
sow_dic(test2,vocabularioEjemplo)


{'selección': 1, 'autor': 1, 'exposición': 1, 'poesía': 1}

**Las palabras del vocabulario que no aparecen en el segundo texto (*gol*, *jugador*, *club*, etc.) no están en el diccionario, por lo que la representación es *sparse*.**

**Ejercicio 2**. Programar una función que, dado un dataset de textos anotados y un vocabulario, construya un diccionario en el que a cada etiqueta se le asocie una lista con los sows correspondientes a los textos de dataset de entrada.

In [ ]:
def dataSetSows(dataSet, vocabulario):
    dicres = dict()
    for etiqueta in dataSet:
        dicres[etiqueta] = []
        for texto in dataSet[etiqueta]:
            sow_represen= sow_dic(texto, vocabulario)
            dicres[etiqueta].append(sow_represen)

    return dicres

**La función *dataSetSows()* tiene dos argumentos: un dataset textual etiquetado (que es un diccionario en el que las claves son las etiquetas y los valores, listas de textos pertenecientes a esa categoría) y un vocabulario.
Primero se crea un diccionario vacío, y un bucle *for* recorre cada etiqueta en el dataset; se crea, acontinuación, una lista vacía en el diccionario *dicres* para almacenar los textos. Con el siguiente bucle a cada texto perteneciente a la categoría se le aplica la función *sow_dic()*, para representarlo como un sow  en forma de diccionario. Con *dicres[etiqueta].append(sow_represen)* se añade la representación Sow del texto correspondiente a la lista de representaciones de su misma etiqueta en el diccionario *dicres*.
De este modo, se devuelve un dicciconario en el que cada clase se asocia con una lista de *sows*.**

In [ ]:
dSSEjemplo = dataSetSows(dataSetEjemplo,vocabularioEjemplo) # dSS = dataSetSows

In [ ]:
dSSEjemplo

{'deportes': [{'gol': 1, 'selección': 1, 'jugador': 1},
  {'selección': 1, 'jugador': 1, 'club': 1},
  {'selección': 1, 'plata': 1, 'oro': 1}],
 'arte': [{'autor': 1},
  {'selección': 1, 'exposición': 1, 'resistencia': 1},
  {'poesía': 1}]}

**En esta salida se observa que el dataset es ahora un diccionario con claves las etiquetas y valores las listas de representaciones dispersas de sows, o sea, de diccionarios en los que la clave es la palabra del vocabulario que aparece en el texto y el valor, uno, que indica su presencia.**

In [ ]:
dSSEjemplo["arte"]

[{'autor': 1},
 {'selección': 1, 'exposición': 1, 'resistencia': 1},
 {'poesía': 1}]

In [ ]:
dSSEjemplo["deportes"]

[{'gol': 1, 'selección': 1, 'jugador': 1},
 {'selección': 1, 'jugador': 1, 'club': 1},
 {'selección': 1, 'plata': 1, 'oro': 1}]

**Rápidamente se observa que la palabra *selección* no será distintiva para clasificar adecuadamente los textos.**

**Ejercicio 3**. Programar una función que, dada una etiqueta y un dataSet anotado (puede ser de textos o de sows), devuelva el número de ejemplos de esa etiqueta que hay en el dataSet (una línea de código). [En todo lo que sigue etiqueta, clase y categoría se utilizarán indistintamente.]



In [ ]:
def numeroEjemplosClase (clase, dataSet):
    return len(dataSet[clase])

**La salida de esta función puede representarse como m(c), el número de ejemplos, es decir, de textos representados como cadenas de *strings* en la categoría c. Para ello aplicamos la función *len(dataSet[clase])*, en la que *dataSet* es un diccionario en el que cada clave representa una categoría y su valor asociado es una lista de ejemplos (textos). Con *len*, se cuenta cuántos ejemplos de cada categoría hay en la lista.**

In [ ]:
numeroEjemplosClase("deportes",dSSEjemplo)

3

In [ ]:
numeroEjemplosClase("arte",dataSetEjemplo)

3

**Se observa, como ya sabíamos que en *dataSetEjemplo* hay tres textos asociados a cada categoría: *arte y deporte*.**

Una ventaja (también se puede considerar una fuente de inseguridad y de errores) del sistema de tipado débil de Python es que una función puede trabajar con argumentos de distinto tipo (como son los dataSet de textos o de sows) siempre que "superficialmente" sean similares:

In [ ]:
numeroEjemplosClase("deportes",dSSEjemplo)

3

**Dado que ambos datasets son diccionarios, aunque el contenido de las listas sea diferente (textos vs. diccionarios sow), la función los considera argumentos del mismo tipo.**

**Ejercicio 4**. Programar una función que, dado un dataSet anotado (puede ser de textos o de sows), devuelva el número de ejemplos del dataSet (apoyarse en la función anterior).

In [ ]:
def numeroEjemplos(dataSet):
    total=0
    for clase in dataSet:
        total+=numeroEjemplosClase(clase,dataSet)
    return total

**La función *numeroEjemplos(dataSet)* se basa en la función *numeroEjemplosClase(clase, texto)*, pues esta devuelve el número de ejemplos (m), o sea de textos, de esa etiqueta que hay en el dataset. De este modo, inicándose el total en cero y con un bucle *for* que recorre cada clase del dataset, se suma el total de textos etiquetados que hay.**

In [ ]:
numeroEjemplos(dataSetEjemplo)

6

In [ ]:
numeroEjemplos(dSSEjemplo)

6

**La función está bien diseñada porque devuelve el resultado correcto, tanto para la versión textual como para la de *sows.***

**Ejercicio 5**. Apoyándose en las dos funciones anteriores, programar una función que, dada una etiqueta y un dataSet, devuelva la probabilidad de que uno de los textos anotados sea de la categoría del argumento (una línea de código).

In [ ]:
def probabilidadClase (clase,dataSet):
    return numeroEjemplosClase (clase, dataSet)/numeroEjemplos(dataSet)

**Esta función devuelve P(c), la probabilidad a priori de la categoría c: P(c)=m(c)/m, calculada con las funciones anteriores (número de ejemplos en la categoría c) / (número total de ejemplos en el dataset).
De este modo, se siguen encadenando funciones previamente programadas y operando entre ellas.**

In [ ]:
probabilidadClase("arte",dataSetEjemplo)

0.5

In [ ]:
probabilidadClase("deportes",dSSEjemplo)

0.5

En un caso como el anterior decimos que el dataSet está equilibrado (o balanceado): mismo número de ejemplos anotados en cada clase.

**Ejercicio 6**. Programar una función que, dada una cadena, una clase y un dataSet de sows, calcule el número de veces en que la cadena aparece en ejemplo de la clase.



In [ ]:
def numeroPalabraEnClase(cadena, clase, dataSetSows):
    contador = 0
    for sow in dataSetSows[clase]:
        if cadena in sow:
            contador += 1
    return contador

**La función tiene tres argumentos: una cadena, una clase y un dataset representado en formato Sow. Su propósito es contar en cuántos textos de la categoría dada aparece una cadena. Para ello, se inicia un contador en cero y, mediante un bucle, se recorre cada sow vinculado a una categoría. Si la cadena está presente en el sow, se incrementa el contador en uno. La función devuelve el número de veces en que la palabra aparece en ejemplos de la clase c, m(c,w) (teniendo en cuenta que trabajamos con sows, que solo contabilizan presencia/ausencia) que es necesaria para calcular la probabilidad condicionada de w dada
c (P(w|c) = m(c,w) / m(c), en la que  m(c) es el número total de textos en la categoría c.**

In [ ]:
numeroPalabraEnClase ("gol","deportes",dSSEjemplo)

1

In [ ]:
numeroPalabraEnClase ("selección","deportes",dSSEjemplo)

3

In [ ]:
numeroPalabraEnClase ("gol","arte",dSSEjemplo)

0

In [ ]:
numeroPalabraEnClase ("selección","arte",dSSEjemplo)

1

***Gol*, por ejemplo, no aparece en nigún texto de arte y en uno de deportes y *selección* aparece en las dos categorías, con lo que, no es discriminante.**

**Ejercicio 7**. Programar una función que, dada una cadena, una clase y un dataSet de sows, calcule la probabilidad de la cadena condicionada a que esté en la clase (una línea de código).


In [ ]:
def probabilidadPalabraEnClase(cadena,clase,dataSetSows):
    return numeroPalabraEnClase(cadena,clase,dataSetSows)/numeroEjemplosClase(clase,dataSetSows)

**Esta función calcula P(w | c) = m(w, c) /m(c), es decir, la probabilidad condicional de que aparezca una palabra, dada una clase. Su cálculo consiste en la divión de las funciones numeroPalabraEnClase  y numeroEjemplosClase.**


In [ ]:
probabilidadPalabraEnClase("gol","deportes",dSSEjemplo)

0.3333333333333333

In [ ]:
probabilidadPalabraEnClase("selección","deportes",dSSEjemplo)

1.0

In [ ]:
probabilidadPalabraEnClase("gol","arte",dSSEjemplo)

0.0

In [ ]:
probabilidadPalabraEnClase("poesía","deportes",dSSEjemplo)

0.0

**Ejercicio 8**. Programar una función que, dado un texto, un vocabulario, una clase y un dataSet de sows, calcule la probabilidad del texto condicionada a que esté en la clase (hay que pasar en primer lugar el texto de entrada a un sow).

In [ ]:
def probabilidadTextoEnClase(texto, vocabulario, clase, dataSetSows):
    sow = sow_dic(texto, vocabulario)
    probabTotal = 1
    for palabra in vocabulario:
        probabPalabra = probabilidadPalabraEnClase(palabra, clase, dataSetSows)

        if palabra in sow:
            probabTotal *= probabPalabra
        else:
            probabTotal *= (1 - probabPalabra)

    return probabTotal

**La función tiene cuatro argumentos: un texto, un vocabulario, una clase y un dataset representado en formato sow. Su propósito es calcular la probabilidad condicionada del texto dada la clase, P(sow | c).
Para ello, el texto se transforma en su representación sow mediante la función *sow_dic(texto, vocabulario)*, que crea un diccionario disperso. En este diccionario, las claves son las palabras presentes en el texto y sus valores indican su presencia con un 1.
La probabilidad total se iguala a 1 y, luego, se recorre el vocabulario. Para cada palabra, se obtiene P(w | c), la probabilidad condicionada de la palabra dada la clase. Si la palabra está presente en el texto, se multiplica por P(w | c), en caso contrario, se multiplica por 1 - P(w | c). Este cálculo sigue el modelo de Bernoulli para el método de Naive Bayes, en el que la probabilidad del texto en una clase se obtiene como el producto de las probabilidades individuales de sus palabras: P(sow | c) = ∏ P(w | c), para cada palabra en el vocabulario.**





In [ ]:
sow_dic(test1,vocabularioEjemplo)

{'gol': 1, 'selección': 1, 'jugador': 1}

In [ ]:
probabilidadTextoEnClase(test1,vocabularioEjemplo,"deportes",dSSEjemplo)

0.06584362139917697

In [ ]:
probabilidadTextoEnClase(test2,vocabularioEjemplo,"deportes",dSSEjemplo)

0.0

In [ ]:
probabilidadTextoEnClase(test2,vocabularioEjemplo,"arte",dSSEjemplo)

0.00823045267489712

**Por ejemplo, *test 2* tiene una probabilidad nula de estar etiquetado en *deportes* y muy baja en *arte*, por lo que algunas de sus palabras clave aparecen en los textos de esta categoría, pero no muchas.**

**Ejercicio 9**. Programar una función que, dado un texto, un vocabulario, una clase y un dataSet de sows, calcule la probabilidad de estar en la clase del texto (una línea de código).

In [ ]:
def probabilidadClaseDeTexto(texto,vocabulario,clase,dataSetSows):
    return probabilidadTextoEnClase(texto,vocabulario,clase,dataSetSows)*probabilidadClase(clase,dataSetSows)

**Esta función aplica el Teorema de Bayes para estimar P(c | sow), la probabilidad condicionada de la clase, dado un texto sow, representado en forma de sow, P(c | Sow = (P(Sow | c) * P(c)) / P(Sow). En la función omitimos el denomidador porque es constante para todas las clases. Así, mientras que en el ercicio 8 se miden cuántos rasgos distintivos del ejemplo aparecen en la clase, en el 9 se multiplican estos rasgos por la probabilidad de la clase en el conjunto de clases para calcular la probabilidad de que el ejemplo pertenezca a una determinada categoría.**



In [ ]:
probabilidadClaseDeTexto(test1,vocabularioEjemplo,"deportes",dSSEjemplo)

0.03292181069958849

In [ ]:
probabilidadClaseDeTexto(test1,vocabularioEjemplo,"arte",dSSEjemplo)

0.0

In [ ]:
probabilidadClaseDeTexto(test2,vocabularioEjemplo,"deportes",dSSEjemplo)

0.0

In [ ]:
probabilidadClaseDeTexto(test2,vocabularioEjemplo,"arte",dSSEjemplo)

0.00411522633744856

**Con estos ejemplos se está tratando de clasificar la pertencia a una clase de dos textos sin etiquetar, mediante el Teorema de Bayes. Por ejemplo:
P(deportes|test1)=P(test1|deportes)⋅P(deportes)=
0.22222 *0.5. En los casos en los que el resultado es cero, como en P(arte | test1) o P(deportes | test2), se debe a que ninguna palabra del texto aparece en los textos de la categoría, por lo que el producto es cero.
Finalmente, el modelo clasifica el texto en la categoría con mayor probabilidad. En este caso, *test1* se clasificaría en deportes y *test2* en arte.**

Véase sin embargo este ejemplo:

In [ ]:
test3 = 'La exposición de poesía cuenta con una selección de obras de un autor y jugador catalán'

In [ ]:
probabilidadClaseDeTexto(test3,vocabularioEjemplo,"deportes",dSSEjemplo)

0.0

In [ ]:
probabilidadClaseDeTexto(test3,vocabularioEjemplo,"arte",dSSEjemplo)

0.0

**Este caso es distinto de los anteriores porque ninguna de las palabras de *test3* está en el vocabulario definido. Como consecuencia, la probabilidad condicional P(test3|clase) es cero para todas las categorías y el modelo no podrá clasificarlo.**

**Ejercicio 10**. Explíquese por qué en ese ejemplo "test3" el programa es incapaz de discriminar y *colapsa a cero*. Modifíquese la función "probabilidadPalabraEnClase" para que no devuelva nunca 0 ni 1, sin desvirtuar las probabilidades relativas (*Laplace Smoothing*). Vuélvanse a ejecutar todos los ejemplos para "probabilidadClaseDeTexto" y coméntense los nuevos resultados.

In [ ]:
def probabilidadPalabraEnClase(cadena, clase, dataSetSows):
    return (numeroPalabraEnClase(cadena, clase, dataSetSows) + 1) / (numeroEjemplosClase(clase, dataSetSows) + 2)

In [ ]:
probabilidadClaseDeTexto(test1,vocabularioEjemplo,"deportes",dSSEjemplo)

0.008493465600000001

In [ ]:
probabilidadClaseDeTexto(test2,vocabularioEjemplo,"arte",dSSEjemplo)

0.0025165824000000013

In [ ]:
probabilidadClaseDeTexto(test3,vocabularioEjemplo,"deportes",dSSEjemplo)

0.0001990656

In [ ]:
probabilidadClaseDeTexto(test3,vocabularioEjemplo,"arte",dSSEjemplo)

0.0006291456000000003

In [ ]:
probabilidadClaseDeTexto(test1,vocabularioEjemplo,"arte",dSSEjemplo)

0.0005308416000000001

In [ ]:
probabilidadClaseDeTexto(test2,vocabularioEjemplo,"deportes",dSSEjemplo)

0.00013271040000000002

**La función *probabilidadPalabraEnClase* se ha modificado porque en el modelo Bernoulli (con sows) de Naive Bayes se deben evitar probabilidades de 0 y 1, para que no haya errores en la clasificación. A este método se le denomina
*Laplace Smoothing*.
Dado que con sows solo se atiende a la presencia o ausencia de palabras, la fórmula adecuada, en este caso, es sumar 1 al numerador y 2 al denominador, con lo que se impide que una palabra tenga una probabilidad de cero en una categoría y que, en consecuencia, su ausencia en un texto anule la probabilidad del documento entero.**



**De este modo, en los ejemplos anteriores a la explicación de *Laplace Smoothing*, se observaba que, si una palabra del vocabulario no aparecía en ningún texto etiquetado en una determinada categoría, como por ejemplo sucedía con *test3*, su probabilidad condicional era cero, y, en consecuencia la probabilidad total del texto en esa categoría también era cero, por lo que no podía clasificarse. Tras aplicar el suavizado, dado que se asigna siempre una pequeña probabilidad, se puede clasificar en la categoría más probable, *arte* en este caso.**

Comenzamos ahora el cálculo de *Naive Bayes* con el modelo BoW, el modelo de documento también llamado multinomial. Vamos a ir repitiendo las funciones del modelo Bernouilli, una a una pero adaptándolas al modelo multinomial. En algunos casos las funciones sirven sin modificación, pero, por claridad, mantendremos los nombres de las funciones, pero terminándolos por la letra mayúscula M (de "Multinomial").

**Ejercicio 11**. Retomar de las prácticas anteriores todas las funciones necesarias para poder programar bows (con diccionarios). Escribirlas en una única celda, una tras otra, y ejecutar la celda (si tenéis todo el cuaderno ejecutado, en particular la celda correspondiente a la solución del Ejercicio 1, no os hará falta recargar todo).

In [ ]:
def frecuencia(cadena, listaCadenas):
    contador = 0
    for palabra in listaCadenas:
        if coincide(cadena, palabra):
            contador= contador + 1
    return contador

def bow_dicM (texto, vocabulario):
    bow_dictM = {}
    for cadena in vocabulario:
        bow_dictM[cadena] = frecuencia(cadena, word_tokenize(texto))
    return bow_dictM


**Debido a que en el ejercicio anterior se ha trabajado con sows dispersos que no almacenan ceros y son más eficioentes computacionalmente, vamos a mejorar la definición de *bow_dict* para que también los bows sean dispersos. Así:**

In [ ]:
def bow_dicM(texto, vocabulario):
    bow_dictM = {}
    for cadena in vocabulario:
        freq = frecuencia(cadena, word_tokenize(texto))
        if freq > 0:
            bow_dictM[cadena] = freq
    return bow_dictM

In [ ]:
bow_dicM(test1, vocabularioEjemplo)

{'gol': 1, 'selección': 1, 'jugador': 1}

**Ejercicio 12**. Programar una función que, dado un dataset de textos anotados y un vocabulario, construya un diccionario en el que a cada etiqueta se le asocie una lista con los bows correspondientes a los textos de dataset de entrada.

In [ ]:
def dataSetBowsM(dataSet, vocabulario):
    dicres = dict()
    for etiqueta in dataSet:
        dicres[etiqueta] = []
        for texto in dataSet[etiqueta]:
            bow_represen= bow_dicM(texto, vocabulario)
            dicres[etiqueta].append(bow_represen)

    return dicres

**Esta función es equivalente a *dataSetSows*, pero en lugar de construir un diccionario en el que a cada etiqueta se le asigne el valor, que es la lista de sows, se asignan la lista de bows, para poder aplicar posteriormente el método Naive Bayes multinomial, basado en frecuencias y no en la presencia y ausencia de palabras. Por este motivo, las funciones siguientes, no se describirán detalladamente, si no es necesario, porque ya se han descrito las funciones homólogas del modelo Bernoulli.**

In [ ]:
bow_dicM(test2,vocabularioEjemplo)

{'selección': 1, 'autor': 1, 'exposición': 1, 'poesía': 1}

In [ ]:
dSBEjemplo = dataSetBowsM(dataSetEjemplo,vocabularioEjemplo) # dSB = dataSetBows

In [ ]:
dSBEjemplo

{'deportes': [{'gol': 2, 'selección': 1, 'jugador': 1},
  {'selección': 1, 'jugador': 1, 'club': 1},
  {'selección': 1, 'plata': 1, 'oro': 1}],
 'arte': [{'autor': 2},
  {'selección': 1, 'exposición': 4, 'resistencia': 2},
  {'poesía': 2}]}

In [ ]:
def numeroEjemplosClaseM (clase, dataSet):
  return numeroEjemplosClase(clase, dataSet)

In [ ]:
numeroEjemplosClaseM("deportes",dataSetEjemplo)

3

In [ ]:
numeroEjemplosClaseM("arte",dSBEjemplo)

3

In [ ]:
def numeroEjemplosM(dataSet):
    return numeroEjemplos(dataSet)

In [ ]:
numeroEjemplosM(dataSetEjemplo)

6

In [ ]:
def probabilidadClaseM (clase,dataSet):
  return probabilidadClase (clase,dataSet)

In [ ]:
probabilidadClaseM("arte",dataSetEjemplo)

0.5

In [ ]:
probabilidadClaseM("deportes",dSBEjemplo)

0.5

**Estos ejemplos son los mismos que al principio de la práctica porque las funciones empleadas no dependen del tipo de representación del texto (Sow o Bow)**.

**Ejercicio 13**. Programar una función que, dada una cadena, una clase y un dataSet de bows, calcule el número de veces en que la cadena aparece en ejemplo de la clase.

In [ ]:
def numeroPalabraEnClaseM (cadena,clase,dataSetBows):
    contador=0
    for bow in dataSetBows[clase]:
        if cadena in bow:
            contador += bow[cadena]
    return contador

**La función tiene tres argumentos: una cadena, una clase y un dataset representado en formato Bow. Su propósito es contar cuántas veces aparece la cadena en los textos de la categoría dada. Para ello, se inicia un contador en cero y, mediante un bucle, se recorre cada Bow vinculado a una categoría. Si la cadena está presente en el Bow, se incrementa el contador con el número exacto de veces que aparece en ese texto (bow[cadena]).
La función devuelve el número total de veces en que la palabra aparece en ejemplos de la clase c, m(c,w), teniendo en cuenta que trabajamos con Bows, que contabilizan frecuencias en lugar de presencia o ausencia. Esta función es necesaria para calcular la probabilidad condicional de w dada c, P(w|c) = m(c,w) / m(c), en la que m(c) es el número total de palabras en los textos de la categoría c.**



In [ ]:
numeroPalabraEnClaseM("gol","deportes",dSBEjemplo)

2

In [ ]:
numeroPalabraEnClaseM("selección","deportes",dSBEjemplo)

3

In [ ]:
numeroPalabraEnClaseM("gol","arte",dSBEjemplo)

0

In [ ]:
numeroPalabraEnClaseM ("selección","arte",dSSEjemplo)

1

**En estos ejemplos se observa la diferencia de uno y otro modelo (Bernoulli y multinomial) en *gol*, es decir en valores mayores a uno, que son los no coincidentes**

**Ejercicio 14**. Programar una función que, dada una cadena, una clase y un dataSet de bows, calcule la probabilidad de la cadena condicionada a que esté en la clase (una línea de código).

In [ ]:
def probabilidadPalabraEnClaseM(cadena, clase, dataSetBows):
    return (numeroPalabraEnClaseM(cadena, clase, dataSetBows) + 1) / (numeroEjemplosClaseM(clase, dataSetBows) + 1)

**En el modelo Bernoulli, la probabilidad condicionada de la palabra, dada la clase P(w | c) = m(w, c) /m(c) en el conjunto de sows, se tiene en cuenta no solo si una palabra aparece, sino también si está ausente: si está, se multiplica su probabilidad en esa clase. Si no, se multiplica  1 - P(w | C).
La probabilidad condicional total del texto en una clase se calcula multiplicando estos valores para todas las palabras del vocabulario, lo que se expresa como el producto de P(w | C) para las palabras que aparecen y (1 - P(w | C)) para las palabras que no aparecen. De ahí que sean problemáticas tanto las probabilidades iguales a cero que las iguales a uno.
En el modelo multinomial, en cambio, solo importa la frecuencia con la que aparece una palabra, por lo que si una palabra no está en el texto, no contribuye al cálculo y no importa que su probabilidad sea uno, por lo que para aplicar el suavizado de Laplace basta con sumar uno al numerador y uno al denominador para evitar cero o una solución no definida para los reales.**

In [ ]:
probabilidadPalabraEnClaseM("gol","deportes",dSBEjemplo)

0.75

In [ ]:
probabilidadPalabraEnClaseM("selección","deportes",dSBEjemplo)

1.0

In [ ]:
probabilidadPalabraEnClaseM("gol","arte",dSBEjemplo)

0.25

In [ ]:
probabilidadPalabraEnClaseM("poesía","deportes",dSBEjemplo)

0.25

In [ ]:
probabilidadPalabraEnClaseM("gol","arte",dSBEjemplo)

0.25

**Los valores en este modelo son más altos para aquellas palabras con alta frecuencia en una determianda categoría, pues en el modelo Bernoulli solo importa la presencia de esas palabras.**

**Ejercicio 15**. Programar una función que, dado un texto, un vocabulario, una clase y un dataSet de bows, calcule la probabilidad del texto condicionada a que esté en la clase (pasar en primer lugar el texto de entrada a un bow). [Aquí es donde reside la mayor diferencia entre el modelo Multinomial y el de Bernouilli; todo el resto es muy similar.]

In [ ]:
def probabilidadTextoEnClaseM(texto, vocabulario, clase, dataSetBows):
    bow = bow_dicM(texto, vocabulario)
    probabTotal = 1
    for palabra in bow:
        probabPalabra = probabilidadPalabraEnClaseM(palabra, clase, dataSetBows)
        probabTotal *= probabPalabra ** bow[palabra]

    return probabTotal

**La probabilidad del texto condicionada a que esté en la clase, P (bow | c) en el modelo multinomial se calcula como el producto de las probabilidades condicionales de las palabras, dadas la clase, elevadas a su frecuencia en el texto: P(texto | c) = ∏ P(w| c) ^ N(w, texto).**

**Para ello esta función convierte el texto en su representación bow, mediante la función *bow_dic*, que devuelve un diccionario disperso con la frecuencia de cada palabra en el texto. Luego, se inicia la probabilidad total en uno y se itera cada palabra presente en el bow para obtener la probabilidad condicionada de la palabra, dada la clase, con la función *probabilidadPalabraEnClaseM*. La probabilidad total se obtiene multiplicando la probabilidad condicional de cada palabra en la clase, elevada a la potencia que corresponde a su frecuencia en el texto.**

**La probabilidad del texto condicionada a que esté en la clase, P(bow | c), en el modelo multinomial se calcula como el producto de las probabilidades condicionales de las palabras dadas la clase, elevadas a su frecuencia en el texto: P(texto | c) = ∏ P(w | c) ^ N(w, texto).
Para ello, esta función convierte el texto en su representación bow mediante la función *bow_dic*, que devuelve un diccionario disperso con la frecuencia de cada palabra en el texto. Luego, se inicializa la probabilidad total en uno y se itera sobre cada palabra presente en el bow. Para cada palabra, se obtiene su probabilidad condicional en la clase mediante la función probabilidadPalabraEnClaseM. La probabilidad total se obtiene multiplicando la probabilidad condicional de cada palabra en la clase, elevada a su frecuencia en el texto.**

In [ ]:
probabilidadTextoEnClaseM(test1,vocabularioEjemplo,"deportes",dSBEjemplo)

0.5625

In [ ]:
probabilidadTextoEnClaseM(test1,vocabularioEjemplo,"arte",dSBEjemplo)

0.03125

In [ ]:
probabilidadTextoEnClaseM(test2,vocabularioEjemplo,"deportes",dSBEjemplo)

0.015625

In [ ]:
probabilidadTextoEnClaseM(test2,vocabularioEjemplo,"arte",dSBEjemplo)

0.3515625

**Ejercicio 16**. Programar una función que, dado un texto, un vocabulario, una clase y un dataSet de bows, calcule la probabilidad de estar en la clase del texto (una línea de código).

In [ ]:
def probabilidadClaseDeTextoM(texto, vocabulario, clase, dataSetBows):
    return probabilidadTextoEnClaseM(texto, vocabulario, clase, dataSetBows) * probabilidadClaseM(clase, dataSetBows)

**Esta función aplica el Teorema de Bayes en el modelo multinomial, ya que estimma P(c | bow), la probabilidad condicionada de la clase, dado un texto representado como bow, P(c | bow) = (P(bow | c) * P(c)) / P(bow).**

In [ ]:
probabilidadClaseDeTextoM(test1,vocabularioEjemplo,"deportes",dSBEjemplo)

0.28125

In [ ]:
probabilidadClaseDeTextoM(test1,vocabularioEjemplo,"arte",dSBEjemplo)

0.015625

In [ ]:
probabilidadClaseDeTextoM(test2,vocabularioEjemplo,"deportes",dSBEjemplo)

0.0078125

In [ ]:
probabilidadClaseDeTextoM(test2,vocabularioEjemplo,"arte",dSBEjemplo)

0.17578125

**Las probabilidades en el modelo multinomial son, generalmente, más altas porque se considera la frecuencia de las palabras en la clase, no su prersencia.**

En el ejemplo de test3, sigue colapsando a cero.

**No colapsa a cero porque ya se ha aplicado el suavizado de Laplace previamente.**

In [ ]:
probabilidadClaseDeTextoM(test3,vocabularioEjemplo,"deportes",dSBEjemplo)

0.005859375

In [ ]:
probabilidadClaseDeTextoM(test3,vocabularioEjemplo,"arte",dSBEjemplo)

**Ejercicio 17**. Aplíquese una solución similar (¿sirve la misma?) a la del Ejercicio 10 (*Laplace Smoothing*) re-definiendo la función "probabilidadPalabraEnClaseM" para evitar ese *colapso a cero*. Vuélvanse a ejecutar todos los ejemplos para "probabilidadClaseDeTextoM" y coméntense los nuevos resultados.

**Ya se ha aplicado en el ejercicio 14, por lo que aquí se define la función sin el suavizado para ver el colapso a cero.**

In [ ]:
def probabilidadPalabraEnClaseM(cadena,clase,dataSetBows):
    return numeroPalabraEnClaseM(cadena,clase,dataSetBows)/numeroEjemplosClaseM(clase,dataSetBows)


In [ ]:
probabilidadClaseDeTextoM(test3,vocabularioEjemplo,"deportes",dSBEjemplo)

0.0

#**Ejercicio opcional 1**

Añádense más ejemplos sobre "deportes" y "arte" en el dataSet de ejemplo. Probad la capacidad de predicción con más ejemplos. Jugad con distintos vocabularios, para intentar mejorar los resultados. ¿Se aprecian diferencias de rendimiento entre los modelos de Bernouilli y el Multinomial?

In [ ]:
dataSetEjemplo = {
    "deportes": [
        "Con un segundo gol, el centrocampista del PSG rescata del asedio helvético a la selección, que jugó más de una hora con uno menos por la expulsión de un jugador. El tercer gol fue anulado.",
        "El jugador asegura que es inocente y pide perdón al club por incumplir el rigor de los horarios y no poder ir a la selección.",
        "La plata olímpica y el oro europeo de la selección femenina impulsan la idea de crear futuras competiciones nacionales.",
        "Las tenista española, experta en tierra batida, celebra su clasificación para el torneo de Pasrís.",
        "Las vallas españolas florecen en los Europeos de atletismo en Apeldoorn.",
        "Oriol Cardona, campeón del mundo de esquí de montaña por aplastamiento",
        "El equipo de fútbol inglés se impuso al PSG con 28 disparos contra uno, el desequilibrio más acusado en un encuentro de eliminatorias de la Champions.",
        "MotoGP cambia Michelin por Pirelli y pasa a un suministrador único para todas las categorías",
        "Análisis de ajedrez | Adiós a Spassky (VI)",
        "Una gimnasta universitaria hace historia tras ejecutar el salto de Simone Biles.",
        "El Rey LeBron James hace historia de nuevo al superar los 50.000 puntos en la NBA.",
        "La atleta manchega que llega desde los 200m, reactivo químico para que el relevo español llegue cargado de esperanzas a los Europeos de Apeldoorn.",
        "El Atlético frena el sueño del Granada en la ida de la semifinal de la Copa de la Reina.",

    ],
    "arte": [
        "El autor de Soldados de Salamina se convierte en el quinto escritor español entrevistado en 70 años por la revista neoyorquina que orienta el canon contemporáneo. Se trata de un reconocido autor de novelas y de no ficción.",
        "Una exposición en Berlín rinde homenaje a las mujeres de diversas clases e ideologías que se jugaron la vida para hacer frente al nacionalsocialismo desde la resistencia.",
        "Poetas consagrados y las voces más jóvenes de la poesía española reflexionan sobre la obra más icónica de Rafael Alberti, escrita cuando tenía 22 años y por la que consiguió el Premio Nacional de poesía en 1924.",
        "Un juez ordena a la UNAM cerrar la exposición ‘La venida del señor’ del artista Fabián Cháirez por “vulnerar la libertad religiosa.”",
        "Las luces y sombras de Caravaggio vuelven a Roma con la mayor exposición en décadas.",
        "Una exposición en Berlín rinde homenaje a las mujeres de diversas clases e ideologías.",
        "Mujeres en prisión e ideal feminista: el TNC rescata ‘Marie la Roja’, de Rosa Maria Arquimbau.",
        "El Museo de América expone una faceta poco conocida de este maestro de la fotografía:imágenes, muchas de ellas inéditas.",
        "El artista ha ideado un conjunto de ocho cabezas para denunciar la intolerancia respecto a la inmigración en ‘Entre sueños V3.0′.",
        "Pol Taburet, el veinteañero francés nieto de migrantes que ha logrado que los museos compitan por su obra.",
        "Una pintura de Pieter Brueghel el Joven robada en Polonia en 1974 reaparece en Países Bajos",
        "Un artista estadounidense se dedica desde hace décadas a recopilar listas de la compra anónimas. En lo más mundano halla lo extraordinario.",
        "Las vanguardias artísticas en la colección del Museo Reina Sofía."


    ]
}

In [ ]:
vocabularioEjemploAmpliado = ["gol", "selección", "jugador", "club", "plata", "fotografía", "museo", "obra", "artista", "pintura", "ajedrez", "clasificación", "atleta", "jugador", "campeón" "oro", "autor", "exposición", "poesía", "resistencia"]

In [ ]:
dataSetSows(dataSetEjemplo, vocabularioEjemploAmpliado)

{'deportes': [{'gol': 1, 'selección': 1, 'jugador': 1},
  {'selección': 1, 'jugador': 1, 'club': 1},
  {'selección': 1, 'plata': 1},
  {'clasificación': 1},
  {},
  {},
  {},
  {},
  {'ajedrez': 1},
  {},
  {},
  {'atleta': 1},
  {}],
 'arte': [{'autor': 1},
  {'exposición': 1, 'resistencia': 1},
  {'obra': 1, 'poesía': 1},
  {'artista': 1, 'exposición': 1},
  {'exposición': 1},
  {'exposición': 1},
  {},
  {'fotografía': 1},
  {'artista': 1},
  {'obra': 1},
  {'pintura': 1},
  {'artista': 1},
  {}]}

In [ ]:
dataSetBowsM(dataSetEjemplo, vocabularioEjemploAmpliado)

{'deportes': [{'gol': 2, 'selección': 1, 'jugador': 1},
  {'selección': 1, 'jugador': 1, 'club': 1},
  {'selección': 1, 'plata': 1},
  {'clasificación': 1},
  {},
  {},
  {},
  {},
  {'ajedrez': 1},
  {},
  {},
  {'atleta': 1},
  {}],
 'arte': [{'autor': 2},
  {'exposición': 1, 'resistencia': 1},
  {'obra': 1, 'poesía': 2},
  {'artista': 1, 'exposición': 1},
  {'exposición': 1},
  {'exposición': 1},
  {},
  {'fotografía': 1},
  {'artista': 1},
  {'obra': 1},
  {'pintura': 1},
  {'artista': 1},
  {}]}

In [ ]:
dSSEjemplo = dataSetSows(dataSetEjemplo, vocabularioEjemploAmpliado)
dSBEjemplo = dataSetBowsM(dataSetEjemplo, vocabularioEjemploAmpliado)


In [ ]:
test1 = "Una gimnasta española logra oro en los europeos"
test2 = "La exposición de arte moderno presenta pinturas inéditas"
test3= "El museo expone una exposición de pintura, pintura, pintura y más pintura"


In [ ]:
probabilidadClaseDeTexto(test1, vocabularioEjemploAmpliado, "deportes", dSSEjemplo)

0.053717576813950264

In [ ]:
probabilidadClaseDeTexto(test1, vocabularioEjemploAmpliado, "arte", dSSEjemplo)

0.044764647344958546

In [ ]:
probabilidadClaseDeTexto(test2, vocabularioEjemploAmpliado, "deportes", dSSEjemplo)

0.003836969772425019

In [ ]:
probabilidadClaseDeTexto(test2, vocabularioEjemploAmpliado, "arte", dSSEjemplo)

0.02238232367247927

In [ ]:
probabilidadClaseDeTexto(test3, vocabularioEjemploAmpliado, "deportes", dSSEjemplo)

1.9576376389923562e-05

In [ ]:
probabilidadClaseDeTexto(test3, vocabularioEjemploAmpliado, "arte", dSSEjemplo)

0.0002459596007964755

In [ ]:
probabilidadClaseDeTextoM(test1, vocabularioEjemploAmpliado, "deportes", dSBEjemplo)

0.5

In [ ]:
probabilidadClaseDeTextoM(test1, vocabularioEjemploAmpliado, "arte", dSBEjemplo)

0.5

In [ ]:
probabilidadClaseDeTextoM(test2, vocabularioEjemploAmpliado, "deportes", dSBEjemplo)

0.03571428571428571

In [ ]:
probabilidadClaseDeTextoM(test2, vocabularioEjemploAmpliado, "arte", dSBEjemplo)

0.17857142857142858

In [ ]:
probabilidadClaseDeTextoM(test3, vocabularioEjemploAmpliado, "deportes", dSBEjemplo)

6.640515431495378e-08

In [ ]:
probabilidadClaseDeTextoM(test3, vocabularioEjemploAmpliado, "arte", dSBEjemplo)

5.312412345196303e-06

In [ ]:
vocabulario_mejor = [

    "gol", "jugador", "selección", "club", "atleta", "fútbol", "gimnasta",
    "torneo", "esquí", "motogp", "champions", "nba", "ajedrez", "psg", "clasificación",

    "autor", "obra", "poesía", "exposición", "fotografía", "pintura", "artista",
    "museo", "colección", "vanguardias", "resistencia", "novelas", "caravaggio",
    "reina", "feminista"
]


In [ ]:
dataSetSows(dataSetEjemplo, vocabulario_mejor)

{'deportes': [{'gol': 1, 'jugador': 1, 'selección': 1},
  {'jugador': 1, 'selección': 1, 'club': 1},
  {'selección': 1},
  {'torneo': 1, 'clasificación': 1},
  {},
  {'esquí': 1},
  {'fútbol': 1},
  {},
  {'ajedrez': 1},
  {'gimnasta': 1},
  {},
  {'atleta': 1},
  {}],
 'arte': [{'autor': 1, 'novelas': 1},
  {'exposición': 1, 'resistencia': 1},
  {'obra': 1, 'poesía': 1},
  {'exposición': 1, 'artista': 1},
  {'exposición': 1},
  {'exposición': 1},
  {'feminista': 1},
  {'fotografía': 1},
  {'artista': 1},
  {'obra': 1},
  {'pintura': 1},
  {'artista': 1},
  {'colección': 1, 'vanguardias': 1}]}

In [ ]:
dataSetBowsM(dataSetEjemplo, vocabulario_mejor)

{'deportes': [{'gol': 2, 'jugador': 1, 'selección': 1},
  {'jugador': 1, 'selección': 1, 'club': 1},
  {'selección': 1},
  {'torneo': 1, 'clasificación': 1},
  {},
  {'esquí': 1},
  {'fútbol': 1},
  {},
  {'ajedrez': 1},
  {'gimnasta': 1},
  {},
  {'atleta': 1},
  {}],
 'arte': [{'autor': 2, 'novelas': 1},
  {'exposición': 1, 'resistencia': 1},
  {'obra': 1, 'poesía': 2},
  {'exposición': 1, 'artista': 1},
  {'exposición': 1},
  {'exposición': 1},
  {'feminista': 1},
  {'fotografía': 1},
  {'artista': 1},
  {'obra': 1},
  {'pintura': 1},
  {'artista': 1},
  {'colección': 1, 'vanguardias': 1}]}

In [ ]:
dSSEjemplo = dataSetSows(dataSetEjemplo, vocabulario_mejor)
dSBEjemplo = dataSetBowsM(dataSetEjemplo, vocabulario_mejor)

In [ ]:
probabilidadClaseDeTexto(test1, vocabulario_mejor, "deportes", dSSEjemplo)

0.003355951346002537

In [ ]:
probabilidadClaseDeTexto(test1, vocabulario_mejor, "arte", dSSEjemplo)

0.001112943048419208

In [ ]:
probabilidadClaseDeTexto(test2, vocabulario_mejor, "deportes", dSSEjemplo)

0.0015581202677868917

In [ ]:
probabilidadClaseDeTexto(test2, vocabulario_mejor, "arte", dSSEjemplo)

0.007790601338934458

In [ ]:
probabilidadClaseDeTexto(test3, vocabulario_mejor, "deportes", dSSEjemplo)

7.949593202994348e-06

In [ ]:
probabilidadClaseDeTexto(test3, vocabulario_mejor, "arte", dSSEjemplo)

8.561100372455443e-05

In [ ]:
probabilidadClaseDeTextoM(test1, vocabulario_mejor, "deportes", dSBEjemplo)

0.07142857142857142

In [ ]:
probabilidadClaseDeTextoM(test1, vocabulario_mejor, "arte", dSBEjemplo)

0.03571428571428571

In [ ]:
probabilidadClaseDeTextoM(test2, vocabulario_mejor, "deportes", dSBEjemplo)

0.03571428571428571

In [ ]:
probabilidadClaseDeTextoM(test2, vocabulario_mejor, "arte", dSBEjemplo)

0.17857142857142858

In [ ]:
probabilidadClaseDeTextoM(test3, vocabulario_mejor, "deportes", dSBEjemplo)

6.640515431495378e-08

In [ ]:
probabilidadClaseDeTextoM(test3, vocabulario_mejor, "arte", dSBEjemplo)

5.312412345196304e-06

## **Comentario**

**En este ejercicio se han aplicado los modelos Naive Bayes Bernoulli y Multinomial sobre un dataset ampliado en las categorías *deportes* y *arte*, con el fin de evaluar su rendimiento clasificatorio bajo distintas condiciones léxicas. La tarea se ha desarrollado en dos fases diferenciadas: la primera con un vocabulario reducido, y la segunda con un vocabulario más representativo de cada clase. En ambas se han utilizado los mismos tres textos de prueba: *test1, test2 y test3*.**

###**Fase 1: Vocabulario ampliado general**

**En la primera fase, el texto *test1* ha sido correctamente clasificado por el modelo Bernoulli como deportes, si bien la cercanía entre valores indica una clasificación con confianza limitada con una probabilidad de 0.0537 frente a 0.0448 para la clase *arte*. Sin embargo, el modelo Multinomial ha asignado 0.5 a ambas clases, hecho que refleja una incapacidad de decisión debida a la escasa aportación del vocabulario frente a las características del texto.**

**El *test2* ha sido correctamente clasificado como *arte* por ambos modelos. En Bernoulli, *arte* ha recibido una probabilidad de 0.0224, superior al valor 0.0038 asignado a *deportes*. El modelo Multinomial ha realizado una clasificación con mayor seguridad: 0.1786 de probabilidad para *arte* frente a 0.0357 para *deportes*.**

**El *test3*  ha sido también adecuadamente clasificado como *arte*. En Bernoulli, la probabilidad para esta clase ha sido de 0.000246, notablemente superior al valor 1.96e-5 determinado para *deportes* (aunque ambos valores han sido muy bajos). El modelo Multinomial, en cambio, ha discriminado con mayor nitidez: 5.31e-6 para *arte* y 6.64e-8 para *deportes*, hecho que refleja cómo este modelo aprovecha la repetición de términos como *pintura* para reforzar su decisión.**

###**Fase 2: Vocabulario mejorado por categoría**

**En la segunda fase, con un vocabulario más amplio los modelos han mostrado una mayor diferenciación categórica.**

**En *test1*, Bernoulli ha mantenido su decisión por la categoría *deportes* con 0.00336 frente a 0.00111 para *arte*. Aunque los valores absolutos han descendido, la diferencia ha sido suficiente para la clasificación. El modelo Multinomial ha categorizado también como *deportes*, con una probabilidad de 0.0714 frente a 0.0357 para *arte*, es decir con mayor claridad en la distinción respecto a la fase anterior.**

**En *test2*, tanto el modelo Bernoulli como el Multinomial han mejorado su rendimiento respecto a la fase anterior. Bernoulli ha asignado un valor de 0.00779 a la clase *arte*, frente a 0.00156 a *deportes* y el modelo Multinomial, por su parte, ha asignado 0.1786 a *arte* y 0.0357 a *deportes*, por lo que se refleja una mayor discriminación clasificatoria.**

**En *test3,* Bernoulli ha proporcionado una probabilidad de 8.56e-5 para *arte* y 7.95e-6 para *deportes*, clasificación correcta aunque con baja seguridad. El modelo Multinomial, en cambio, ha sido contundente: 5.31e-6 para *arte* frente a 6.64e-8 para *deportes*.**


###**Conclusión**

**El análisis ha demostrado que el modelo Bernoulli tiende a ser preciso cuando la presencia de ciertos términos es suficiente para inferir la clase, pero ofrece decisiones más débiles cuando las diferencias son mínimas. Frente a ello, el modelo Multinomial ha mostrado una mayor precisión, especialmente en textos con repeticiones o alta densidad semántica, como en el caso de *test3*, en el que los valores extremadamente bajos asignados por los modelos con ambos vocabularios se deben al uso del suavizado de Laplace, que evita probabilidades nulas en presencia de términos escasamente frecuentes en el dataset.**

**Por último, puede destacarse cómo un vocabulario amplio, diseñado en función del dominio temático, no solo mejora el rendimiento predictivo de los dos modelos, sino que propicia clasificaciones más discriminativas.**



**Ejercicio opcional 2**. Elíjase un caso de estudio más realista, con más etiquetas/clases, con más datos de entrenamiento y repítase el ejercicio anterior.

**Ejercicio opcional 3**. Prográmese el algoritmo de clasificación 1NN (KNN con K = 1) y compárense los resultados en todos los ejemplos anteriores con *Naive Bayes* en sus dos variantes.